lib/streamlit/cli.py

```diff
@@ -200,6 +204,11 @@ def _get_command_line_as_string() -> Optional[str]:
     parent = click.get_current_context().parent
     if parent is None:
         return None
+
+    # Assert that the program name we see here is `streamlit`, even if we ran
+    # streamlit some other way than `streamlit run`.
+    assert parent.command_path == "streamlit"
+
```

Comment:
What are circumstances where this can get triggered? Could user code accidentally trigger it, or is it just a regression sanity check for ourselves internally? (If user code could trigger it probably should have an error message; else it's fine!)

Reply:
I'm ~99% sure that user code shouldn't be able to trigger this, and it's intended to be an internal sanity check used by the revised smoke tests.

I have seen some instances of people importing `streamlit.bootstrap` and calling `bootstrap.run()` to start streamlit scripts, but that shouldn't run into this issue / people would probably have to have been pretty creative to find ways of invoking streamlit scripts so that this assert fails.

Reply:
FYI, I just run into this assertion today by running:

```shell
python -m streamlit.cli hello
```

I am just noting it here - perhaps this is expected

Reply:
Hey @nrontsis, thanks for pointing this out.

I think we'll have to look into how many people are using `python -m streamlit.cli <command>` to run streamlit, but this isn't a way we expect people to run streamlit (looking at the `git blame`, it looks like it was something added very early in streamlit's lifespan when the cli was first being built that was left alone until now).

It should be pretty easy to avoid this issue: running `python -m streamlit hello` is equivalent, and we officially support this method so provide some guarantees around the stability of it

Reply:
I had someone I work with refer to https://discuss.streamlit.io/t/can-i-use-streamlit-within-a-pycharm-development-environment/4837/4 and https://stackoverflow.com/a/60172283 to run/debug Streamlit through Pycharm configuration, I'm thinking some will come from there...

I'll have to try reusing PyCharm some day in the week, but maybe you have a Jetbrains user that can quickly try and we can definitely document it in this forum issue?

Have a nice day
Fanilo

Reply:
Thanks for the help @andfanilo!

Hm, this may be an interesting unintended consequence to deal with as I believe we were unaware that running `python -m streamlit.cli <command>` is a relatively common pattern within the community.

@tconkling replied to a similar question asked in the forums with an "official" way of running streamlit in PyCharm: https://discuss.streamlit.io/t/version-1-5-0/21455/11

If there are enough people that rely on this behavior, then I'd be okay with relaxing these asserts and even adding some tests to ensure that this doesn't break going forward, but I'd guess this will need a small amount of internal discussion since I don't believe we ever meant for this way of running streamlit to be commonly used

Reply:
Just want to chime in - if you need another data point - that I am also using the `streamlit.cli` method for pycharm configuration:

![image](https://user-images.githubusercontent.com/7430677/151942154-918945c8-f34b-4d44-94a6-e16238f1664f.png)

Reply:
Yeah the invocation I mentioned was also via PyCharm. Following @vdonato's link avoids the assertion, but I wonder if `streamlit.cli` is still supported? If not I can try to open a PR to remove it, if that would be useful.

Reply:
Thanks for the additional data point @nicornk.

@nrontsis: we're still discussing this internally as it seems like this issue is an instance of [Hyrum's law](https://www.hyrumslaw.com/) 🙃 

On one hand, I don't believe we ever intended for this to be a common/supported pattern, so there's some justification in leaving things as-is (and also entirely removing `streamlit.cli` being runnable), especially since an easy way to get to a supported method is to simply remove the `.cli` in the module name.

On the other hand, it seems like this ended up being a very commonly discovered way of running streamlit in PyCharm/VSCode/etc, so there's an argument for just making it supported to avoid disrupting people that have yet to upgrade to v1.5

Reply:
Following up on this: after some discussion, we decided to keep things as-is since:
* it feels weird adding official support to something that essentially only worked accidentally, so we'd prefer not to promote undocumented behavior to an officially supported way of invoking streamlit if there's an easy fix
* an easy fix is to remove the `.cli` from `python -m streamlit.cli <command>` since we do officially support running streamlit via `python -m streamlit <command>`

Sorry for the trouble this may have caused! I just merged https://github.com/streamlit/streamlit/pull/4359 so that people running into this have a better chance of being able to diagnose and solve this problem on their own, but on our end we're also
1. looking into adding official documentation ways to use streamlit with common IDEs/editors (PyCharm/VSCode/etc) as we concluded that part of the reason this happened is because we currently don't have official docs on how to set this up
2. planning on posting on the StackOverflow question / contacting the author of the recommended answer to point out that `python -m streamlit <command>` is the officially blessed method so that the answer can be corrected (I'll be doing this at some point tomorrow)

Reply:
@vdonato Thank you for these efforts. I switched from `streamlit.cli` to `streamlit` in my Pycharm Configuration and verified that running and debugging still works as before.
Appreciate the doc updates for newcomers!